### PR TITLE
CB-13052 Validation of the feature is only allowed on the Native API …

### PR DESCRIPTION
**Pull request by doktoric**
…based AWS variant. Now we only accept multiaz request for AWS_NATIVE

See detailed description in the commit message.

### CODE REVIEW

File: auth-connector/src/main/java/com/sequenceiq/cloudbreak/auth/altus/model/Entitlement.java [hunk 1]
```diff
@@ -7,6 +7,7 @@
     DATAHUB_DEFAULT_SCALING,
     CDP_AZURE,
     CDP_GCP,
+    CDP_AWS_NATIVE,
```
Comment (lajosrodek) [inline auth-connector/src/main/java/com/sequenceiq/cloudbreak/auth/altus/model/Entitlement.java:10]: I suggest adding `_CB` after `CDP` to make it clear that it solely influences CB but not experiences. As I see, the UMS change is still unsubmitted, so we can freely rename this now. 😄

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import com.sequenceiq.cloudbreak.domain.stack.instance.network.InstanceGroupNetwork;
+import com.sequenceiq.cloudbreak.validation.ValidationResult;
+
+@Component
+public class MultiAzValidator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MultiAzValidator.class);
+
+    @Value("${cb.multiaz.supported.variant:AWS_NATIVE}")
+    private Set<String> suporterMultiAzVariants;
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java:30]: Rename: `suporter` -> `supported`

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        if (suporterMultiAzVariants.isEmpty()) {
+            suporterMultiAzVariants = Set.of(AWS_NATIVE_VARIANT.variant().value());
+        }
+    }
+
+    public void validateMultiAzForStack(
+        String variant,
+        Iterable<InstanceGroup> instanceGroups,
+        ValidationResult.ValidationResultBuilder validationBuilder) {
+        Set<String> allSubnetId = collectSubnetIds(instanceGroups);
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java:43]: Nit: rename variable to `allSubnetIds`

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        if (allSubnetId.size() > 1 && !Strings.isNullOrEmpty(variant)) {
+            if (!suporterMultiAzVariants.contains(variant)) {
+                validationBuilder.error(String.format("Multiple Availability zone is not support for %s variant", variant));
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java:46]: Nit:
* `zone` -> `Zone`
* potentially insert `feature` after `Zone`
* `support` -> `supported`

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            }
+        }
+    }
+
+    private Set<String> collectSubnetIds(Iterable<InstanceGroup> instanceGroups) {
+        Set<String> allSubnetId = new HashSet<>();
+        for (InstanceGroup instanceGroup : instanceGroups) {
+            InstanceGroupNetwork instanceGroupNetwork = instanceGroup.getInstanceGroupNetwork();
+            Json attributes = instanceGroupNetwork.getAttributes();
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java:55]: May `instanceGroupNetwork` ever be `null` (e.g. when called for a YCloud stack)?

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        Set<String> allSubnetId = collectSubnetIds(instanceGroups);
+        if (allSubnetId.size() > 1 && !Strings.isNullOrEmpty(variant)) {
+            if (!suporterMultiAzVariants.contains(variant)) {
+                validationBuilder.error(String.format("Multiple Availability zone is not support for %s variant", variant));
+            }
+        }
+    }
+
+    private Set<String> collectSubnetIds(Iterable<InstanceGroup> instanceGroups) {
+        Set<String> allSubnetId = new HashSet<>();
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java:52]: Nit: rename variable to `allSubnetIds`

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidator.java [hunk 7]
```diff
@@ -36,12 +40,28 @@
...
         validateTemplates(subject, validationBuilder);
         validateEncryptionKey(subject, validationBuilder);
+        validateVariant(subject, validationBuilder);
+    }
+
+    private void validateVariant(Stack source, ValidationResultBuilder validationBuilder) {
+        String variant = source.getPlatformVariant();
+        boolean awsNativeEnabled = entitlementService.awsNativeEnabled(source.getCreator().getTenant().getName());
+        if (AWS_NATIVE_VARIANT.variant().value().equals(variant)) {
+            if (awsNativeEnabled) {
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidator.java:59]: The two conditions can be merged into a single statement, unless some further checks are also needed inside the outer `if` block.

Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidator.java:59]: Has to be negated, so `!awsNativeEnabled`

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidator.java [hunk 8]
```diff
@@ -36,12 +40,28 @@
...
+                validationBuilder.error(String.format("%s entitlement was not granted to your tenant. "
+                        + "Please Contact with Cloudera support to request it.",
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidator.java:61]: Nit: `contact with Cloudera support` or `get in contact with Cloudera support`

File: core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidatorTest.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    @Test
+    public void testAwsWhenOneSubnetUsedShouldNotCauseAnyValidationError() {
+        String variant = "AWS";
+        Set<InstanceGroup> instanceGroups = Set.of(
+                instanceGroup(Set.of("subnet-123", "subnet-145")),
+                instanceGroup(Set.of("subnet-123"))
+        );
+        underTest.validateMultiAzForStack(variant, instanceGroups, builder);
+        Mockito.verify(builder, Mockito.times(1)).error(anyString());
+    }
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidatorTest.java:70]: The body seems to be a copy of `testAwsWhenMultipleSubnetUsedShouldCauseAnyValidationError`, so the name does not reflect the expectation.

File: core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidatorTest.java [hunk 10]
```diff
@@ -4,6 +4,8 @@
 import static com.sequenceiq.cloudbreak.validation.ValidationResult.State.VALID;
 import static org.junit.Assert.assertEquals;
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidatorTest.java:5]: Optional: migrate to JUnit 5.

File: core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidatorTest.java [hunk 11]
```diff
@@ -30,6 +33,8 @@
 import com.sequenceiq.cloudbreak.service.rdsconfig.RdsConfigService;
 import com.sequenceiq.cloudbreak.structuredevent.CloudbreakRestRequestThreadLocalService;
 import com.sequenceiq.cloudbreak.validation.ValidationResult;
+import com.sequenceiq.cloudbreak.workspace.model.Tenant;
+import com.sequenceiq.cloudbreak.workspace.model.User;
 
 @RunWith(MockitoJUnitRunner.class)
 public class StackValidatorTest extends StackRequestValidatorTestBase {
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidatorTest.java:40]: Maybe I overlooked it but could not find a test case checking for the failure condition of `StackValidator.validateVariant()`.

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java (hunk 5 above)
Comment (doktoric) [reply core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java:55]: not. really

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+import com.sequenceiq.cloudbreak.domain.stack.instance.network.InstanceGroupNetwork;
+import com.sequenceiq.cloudbreak.validation.ValidationResult;
+
+@Component
+public class MultiAzValidator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MultiAzValidator.class);
+
+    @Value("${cb.multiaz.supported.variant:AWS_NATIVE}")
+    private Set<String> suportedMultiAzVariants;
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidator.java:30]: Typo: suported -> supported

File: core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidatorTest.java (hunk 9 above)
Comment (lajosrodek) [reply core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/network/MultiAzValidatorTest.java:70]: Ping, `testAwsWhenOneSubnetUsedShouldNotCauseAnyValidationError()` and `testAwsWhenMultipleSubnetUsedShouldCauseValidationError()` still have the exact same logic.

File: core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidatorTest.java [hunk 13]
```diff
@@ -121,10 +136,48 @@ public void testWithPositiveRootVolumeSize() {
...
         ValidationResult.ValidationResultBuilder builder = new ValidationResult.ValidationResultBuilder();
 
+        entitlementMock(stack, false);
+
+        underTest.validate(stack, builder);
+        assertEquals(VALID, builder.build().getState());
+    }
+
+    @Test
+    public void testWithPositiveVariantWhenEntitlementEnabled() {
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidatorTest.java:146]: Nit: Rename to `testWithNativeVariantWhenEntitlementEnabled()`

File: core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidatorTest.java [hunk 14]
```diff
@@ -121,10 +136,48 @@ public void testWithPositiveRootVolumeSize() {
...
+        ValidationResult.ValidationResultBuilder builder = new ValidationResult.ValidationResultBuilder();
+
+        entitlementMock(stack, true);
+
         underTest.validate(stack, builder);
         assertEquals(VALID, builder.build().getState());
     }
 
+    @Test
+    public void testWithNegativeVariantWhenEntitlementEnabled() {
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/controller/validation/stack/StackValidatorTest.java:158]: Nit: Rename to `testWithNativeVariantWhenEntitlementDisabled()`

**Review by lajosrodek — APPROVED** [review]
Thanks for the fixes, LGTM

**Pull request merged**